refactor(reports) Refactor open debtors

sfount opened the pull request:
This commit introduces a simpler query for determining open debts.
It sums the total debts for any given source and groups by an entity
uuid - transactions not referencing a debtor are ignored.

sfount committed:
refact(reports) Add optimised query for open debts

This commit introduces a simpler query for determining open debts.
It sums the total debts for any given source and groups by an entity
uuid - transactions not referencing a debtor are ignored.

sfount committed:
refactor(reports) Update server API for debt query

This commit refactors the servers expected attributes from the
request query factoring in new complex vs simple debtor reports
as well as a date limit. The code is seperated into smaller logical
methods.wq

jniles committed:
refactor(reports): document and clean open debtors

This commit introduces extensive documentation to the top of the open
debtors report server module explaining the options.  It also makes sure
that the server code passes lint and/or documenting comments are present
to track future changes that need to be made.

jniles committed:
feat(reports): extra data toggles for Open Debtors

This commit implements two extra toggles for the Open Debtors report:
 1. Show Unverified Transactions - includes the Posting Journal in
 query.
 2. Show Columns for Last Dates of Invoicing and Payment - includes
 extra columns for the last payment date and last invoice date.

This is part of the refactor of the Open Debtors report.  The date limit
still remains to be implemented.

jniles committed:
feat(reports): enable date limit in Open Debtors

This commit adds in the date limit option for the Open Debtors report.
The date limit allows a user to only scan records newer than a certain
date.

jniles committed:
fix(reports): Open Debtors respects order param

This commit fixes a bug with the Open Debtors report where the report
did not respect the ordering parameter.  It also refactors the ordering
to have more succinct code as well as a clearer select by combining last
invoice and last payment into a simpler "last transaction".

jniles committed:
test(reports): update Open Debtors e2e tests

This commit updates the end to end tests to work with the new report
layout.

DedrickEnc commented on client/src/i18n/fr/report.json:
Should read : Ce rapport affiche les entrées et sorties de la trésorerie (caisse) par service.

DedrickEnc commented on client/src/i18n/fr/report.json:
```diff
@@ -32,19 +34,24 @@
+    "OPEN_DEBTORS": {
+      "TITLE":"Débiteurs endettés",
+      "TREE":"Dettes des Débiteurs",
+      "DESCRIPTION": "Ce rapport montre les débiteurs qui ont une dette avec l'institution et donne un certain contrôle sur l'ordre de tri. Pour une analyse avancée, le rapport peut ajouter la dernière date de facturation et la dernière date de paiement, ainsi que limiter la portée de la recherche par une date.",
```
should read : Ce rapport liste les débiteurs et leurs dettes envers l'institution. Permet de trier les données à lister ainsi que de voir la dernière date de facturation et de paiement pour une analyse assez poussée. Les donnees peuvent être filtrées par date aussi.

DedrickEnc commented on client/src/i18n/fr/report.json:
```diff
@@ -32,19 +34,24 @@
+      "SHOW_DETAILED_VIEW" : "Inclure les colonnes pour les dernières dates de facturation et de paiement",
+      "SHOW_DETAILED_VIEW_HELP" : "L'activation de cette option comprendra la dernière date à laquelle le débiteur a payé et la dernière date où ils ont été facturés. Ceci est utile pour comprendre quand ils étaient à l'établissement. Cependant, cela rendra la requête plus lent.",
+      "SHOW_UNVERIFIED_TRANSACTIONS" : "Inclure les transactions non vérifiées",
+      "SHOW_UNVERIFIED_TRANSACTIONS_HELP" : "L'activation de cette option comprend les enregistrements du Journal de Saisie qui n'ont pas été vérifiés par le comptable ainsi que les enregistrements du Grand Livre que le comptable a vérifié. Cette vue doit être traitée comme une projection jusqu'à ce que le comptable ait approuvé les transactions du journal et les a posté vers le Grand Livre."
```
Should read : L'activation de cette option inclut les données du Journal (données non verifiées) et celles du Grand Livre (données vérifiées). Ce rapport doit être consideré comme une simple projection jusqu'à ce que les données du journal soient postées.

jniles replied:
This is much better :+1:

DedrickEnc commented on client/src/i18n/fr/report.json:
```diff
@@ -32,19 +34,24 @@
-      "LAST_PAYMENT": "Date du dernier paiement",
-      "PATIENT_NAME": "Nom du patient",
-      "TOTAL_DEBT": "Total des dettes"
+      "LAST_TRANSACTION": "Date de la Dernière Transaction",
```
You can also avoid arcticles and say : Date Dernière Transaction

DedrickEnc reviewed (dismissed):
@sfount
Your french translation are good, I have just help you making them short and expressive at the same time.
Feel free to omit articles if you want e.g ```Nom Patient``` instead of ```Nom du Patient```.

jniles committed:
fix(i18n): improve Open Debtors FR translations

This commit improves the translation code following the recommendations
of @dedrickenc.

sfount commented on server/controllers/finance/reports/debtors/openDebtors.handlebars:
```diff
@@ -19,19 +20,19 @@
-              <td>{{date debtor.lastInvoiceDate}}</td>
-              <td>{{date debtor.lastPaymentDate}}</td>
-              <td class="text-right">{{currency debtor.debt ../metadata.enterprise.currency_id}}</td>
+              {{#if ../details.showDetailedView}}<td>{{date debtor.lastInvoiceDate}}</td>{{/if}}
```
I think the SQL now uses `latestInvoiceDate` and `latestCashDate`. These references should either be updated or the SQL should be changed to match this interface. 

As it is now I believe this default to the current date of the report as it will be rendering a `new Date(null)`.

jniles replied:
Ah, I hadn't touched the handlebars.  I'll take a look.

sfount commented on server/controllers/finance/reports/debtors/openDebtors.js:
```diff
@@ -35,112 +62,114 @@ function build(req, res, next) {
-    .then(result => res.set(result.headers).send(result.report))
-    .catch(next)
-    .done();
+const DEFAULT_ORDER = 'debt-desc';
```
A default order is set on the server but if nothing is set on the client it will use this as default. Potentially it would be a good user experience to set this default ordering on the client as well so that the user has an idea of which ordering will be used even if they don't select an order.

sfount commented on client/src/modules/reports/generate/open_debtors/open_debtors.html:
```diff
@@ -23,36 +23,101 @@ <h3 class="text-capitalize" translate>REPORT.OPEN_DEBTORS.TITLE</h3>
+                FORM.LABELS.ORDER_BY
               </label>
 
               <ui-select
```
This report has at most 6 options for ordering. Do you think we could use a select here? The benefits of a select are cross browser (especially mobile) functionality and simplicity of use.

jniles replied:
I've opted for `enabled-search="false"` on the ui-select.  It may not be perfect, but it allows the unicode characters to show through.

In the future, we could do any of these things:
 1. Write an CSS class to be applied to `:before` a `select > option` element that will put either a unicode caret up or caret down.
 2. Use text ("first", "last", "largest", "smallest", etc...)  These would need to be standardized across the entire application.

sfount commented on client/src/modules/reports/generate/open_debtors/open_debtors.html:
```diff
@@ -23,36 +23,101 @@ <h3 class="text-capitalize" translate>REPORT.OPEN_DEBTORS.TITLE</h3>
+                <ui-select-match placeholder="{{ 'FORM.LABELS.ORDER_BY' | translate }}">
+                  <span class="text-capitalize">
+                    <i class="fa" ng-class="{
+                      'fa-caret-down': $select.selected.direction === ReportConfigCtrl.DESC,
```
👍 This is a very clear way of showing ordering.

sfount reviewed (commented):
I have tested the user facing side of this PR and it looks great. More detailed descriptions should help users better understand how they can make the most of reports. 

I think there are one or two things to clean up with this PR - these are fairly small fixes and this should be good to be merged.

jniles committed:
chore(reports): default ordering for Open Debtors

This commit configures the open debtors client to order by default on
the amount of debt descending.
